pkg/aiapiclient/aiapiclient.go

```diff
@@ -29,14 +36,85 @@ func NewAiAPIClient() AiAPIClient {
 	defaultAppInsightsAppID := os.Getenv("APP_INSIGHTS_APP_ID")
 	appInsightsKey := os.Getenv("APP_INSIGHTS_KEY")
 
+	// if no application insights key has been specified, then we will use AD authentication
 	return AiAPIClient{
-		appID:  defaultAppInsightsAppID,
-		appKey: appInsightsKey,
+		appID:           defaultAppInsightsAppID,
+		appKey:          appInsightsKey,
+		useADAuthorizer: appInsightsKey == "",
 	}
 }
 
 // GetMetric calls to API to retrieve a specific metric
-func (ai AiAPIClient) GetMetric(metricInfo MetricRequest) (*MetricsResponse, error) {
+func (ai AiAPIClient) GetMetric(metricInfo MetricRequest) (*insights.MetricsResult, error) {
+	if ai.useADAuthorizer {
+		glog.V(2).Infoln("No application insights key provided - using Azure GO SDK auth.")
+		return getMetricUsingADAuthorizer(ai, metricInfo)
+	}
+
+	glog.V(2).Infoln("Application insights key has been provided - using Application Insights REST API.")
+	return getMetricUsingAPIKey(ai, metricInfo)
+}
+
+func getMetricUsingADAuthorizer(ai AiAPIClient, metricInfo MetricRequest) (*insights.MetricsResult, error) {
+
+	authorizer, err := auth.NewAuthorizerFromEnvironmentWithResource(azureAdResource)
+	if err != nil {
+		glog.Errorf("unable to retrieve an authorizer from environment: %v", err)
+		return nil, err
+	}
+
+	metricsClient := insights.NewMetricsClient()
+	metricsClient.Authorizer = authorizer
+
+	metricsBodyParameter := insights.MetricsPostBodySchemaParameters{
+		Interval: &metricInfo.Interval,
+		Timespan: &metricInfo.Timespan,
+		MetricID: insights.MetricID(metricInfo.MetricName),
+	}
+
+	requestSchemaIdentifier := generateRequestSchemaUniqueIdentifier()
+	metricsBody := []insights.MetricsPostBodySchema{
+		insights.MetricsPostBodySchema{
+			ID:         &requestSchemaIdentifier,
+			Parameters: &metricsBodyParameter,
+		},
+	}
+
+	metricsResultsItem, err := metricsClient.GetMultiple(context.Background(), ai.appID, metricsBody)
+	if err != nil {
+		glog.Errorf("unable to get retrive metric: %v", err)
+		return nil, err
+	}
+
+	// check there is a metric result
+	if len(*metricsResultsItem.Value) == 0 {
+		return nil, errors.New("response from metrics request is empty")
+	}
+
+	// take only the first result (as we ask for a specific metric, there is only one result)
+	metricsResult := (*metricsResultsItem.Value)[0]
+
+	// check the body is not nil
+	if metricsResult.Body == nil {
+		return nil, errors.New("response from metrics request is empty")
+	}
+
+	return metricsResult.Body, nil
+}
+
+func generateRequestSchemaUniqueIdentifier() string {
+	b := make([]byte, 16)
+	_, err := rand.Read(b)
+	if err != nil {
+		log.Fatal(err)
+	}
+	uuid := fmt.Sprintf("%x-%x-%x-%x-%x",
+		b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
+
+	return uuid
+}
+
+func getMetricUsingAPIKey(ai AiAPIClient, metricInfo MetricRequest) (*insights.MetricsResult, error) {
```

Comment:
Do you need the whole AiAPIClient? maybe passe just the AppId instead.

Reply:
this looks like it should probably be a receiver function